Rewrite meta-check.nix type checking logic.

Profpatsch opened the pull request:
Description of the main commit:

```
stdenv/generic/check-meta: rewrite type checking code  …
The previous code used some ad-hoc attribute enumeration to check for missing
fields and did not really type check the attributes, because it used `t.check`
from the module system types, which only does shallow checks for nested types,
e.g. just `builtins.isList` for (listOf strings).

This new version uses `types-simple` to recursively check every attribute,
outputting accurate information for type errors in nested fields, e.g.:

Package ‘hello-2.10’ in … has an invalid meta attrset:
maintainers.1.email should be: string
but is: null

When we set
meta.maintainers [ eelco foobar ];
in `hello/default.nix` and foobar is defined as
{ name = "foobar"; email = null; };
in `maintainers-list.nix`.

The path accurately pins down where the type check failed:
In the `maintainers` field, the second element of the list (counting from 0)
and the field email in that element.

Further work: If an unsupported meta field is used, the type error will print
the complete type of the meta product, which is quite big. The type error should
show a diff of the problematic fields in that case.
This is possible but not yet integrated into the type checking logic.
```

<s>Please go trough the commits one-by-one, since they all add some pretty different changes and are built on each other as single units.</s>
Adding the type library might be a bit controversial, but I hope I have provided enough incentive in the docstrings and the commit messages to persuade you that it is a good addition.

Meta can be checked with [this invocation](https://github.com/NixOS/ofborg/#running-meta-checks-locally) from the ofborg readme.

Profpatsch committed:
lib: add types-simple

A simple type system to check plain nix values.
Please see the module comments for further explanations.

Will be used to correct the buggy meta-field checks and help in sanity-checking
other parts of nixpkgs.

Profpatsch committed:
lib/types-simple: add productOpt and defaults

productOpt: a product type with the possibility for optional fields
defaults: generate (semi-arbitrary) default values for a given type

In preparation for the rewrite of `check-meta.nix`.

Profpatsch committed:
stdenv/generic/check-meta: rewrite type checking code

The previous code used some ad-hoc attribute enumeration to check for missing
fields and did not really type check the attributes, because it used `t.check`
from the module system types, which only does shallow checks for nested types,
e.g. just `builtins.isList` for (listOf strings).

This new version uses `types-simple` to recursively check every attribute,
outputting accurate information for type errors in nested fields, e.g.:

Package ‘hello-2.10’ in … has an invalid meta attrset:
maintainers.1.email should be: string
but is: null

When we set
meta.maintainers [ eelco foobar ];
in `hello/default.nix` and foobar is defined as
{ name = "foobar"; email = null; };
in `maintainers-list.nix`.

The path accurately pins down where the type check failed:
In the `maintainers` field, the second element of the list (counting from 0)
and the field email in that element.

Further work: If an unsupported meta field is used, the type error will print
the complete type of the meta product, which is quite big. The type error should
show a diff of the problematic fields in that case.
This is possible but not yet integrated into the type checking logic.

Profpatsch committed:
check-meta: unify meta.licenses attribute type

The meta type is changed to either a string, a license or a list of licenses,
where license is an attrset with required `fullName`, `shortName` and `free`
fields, and two optional fields.

Every package that does not conform to this union type is fixed; I noticed that
only a few packages used `list of strings`, so I corrected them and removed the
alternative.

Profpatsch committed:
lib/types-simple: add lots more (inline) documentation

Especially document the inner workings of the `checkType` function.

Profpatsch commented:
Okay, with the latest commit I recommend looking at `lib/types-simple.nix` in the overview, since it should be a lot easier to understand with these comments.

aszlig commented on lib/types-simple.nix:
```diff
@@ -0,0 +1,457 @@
+  };
+
+  # the type with exactly one inhabitant
+  unit = mkScalar {
```
Why do we need this type?

Profpatsch replied:
Similar to `void` this doesn’t seem useful at first, but it’s even more elementar than `bool` in a sense; e.g. the `void` you find in languages like `C(++)` is really a `unit`.

A great use-case is for sum types:

```
let mode = sum { off = unit; level = int; };
in checkType mode { off = {}; };
```

which corresponds to the Haskell type:

```
data Mode = Off | Level Int
```

aszlig commented on lib/types-simple.nix:
```diff
@@ -0,0 +1,457 @@
+  # list (attrs unit):
+  #   [ { a = {}; } { b = {}; } ]
+  #   []
+  list = t: mkRecursive {
```
Maybe call this `listOf` so we use the same naming as with NixOS modules.

Profpatsch replied:
I deliberately changed the names to be internally consistent; the module system has `string` and `str`, because `string` did strange concatenation. `listOf` and `attrsOf` have the same history with `attrs` and `list`, which are not able to recursively check their contents.

Compare the correspondence table below.

aszlig commented on lib/types-simple.nix:
```diff
@@ -0,0 +1,457 @@
+  # attrs (attrs string):
+  #  { foo.bar = "hello"; baz.quux = "x"; }
+  #  { x = { y = "wow"; }; }
+  attrs = t: mkRecursive {
```
Same here: `attrsOf`

aszlig commented on lib/types-simple.nix:
```diff
@@ -0,0 +1,457 @@
+          optfs = builtins.attrNames opt;
+          vfs   = builtins.attrNames v;
+      in builtins.isAttrs v &&
+      (if opt == {}
```
I'd use `intersectAttrs` (especially because it's a primop) here instead of converting to list and use `subtractLists`.

Profpatsch replied:
To check for `then reqfs == vfs` in the line below, I’d have to use `builtins.intersectAttrs v req == req`, which has to do a deep equality check for `req` (or `v` if I turn the arguments around).

aszlig commented on lib/types-simple.nix:
```diff
@@ -0,0 +1,457 @@
+  # or even recursiveUpdate if you want to be naughty.
+  # Hint: It’s probably better to use `productOpt` instead,
+  #       but more power to the people.
+  defaults = t:
```
Hm, I don't think this has any more benefit to having to specify the defaults directly, especially if it comes to code readability. You need to look up the default here instead of just following the code you're already looking at.

Profpatsch replied:
I implemented that as a crutch for optional fields originally, but `productOpt` kind of replaced it pretty quickly. It might make sense to remove it again, yes.

aszlig reviewed (commented):
Great work on the type system, although I'm not quite sure whether this is a good idea to use for checking meta, because we already have to fight against eval time.

Apart from that, it could also make sense to replace the type checking for the NixOS module system to use this instead (as it avoids the need to use `tryEval` for checking) on everything but submodules, maybe @nbp could chime in for that as well.

Other than that, I think adding another type system needs to have more eyes on this, so I'd suggest adding this as a separate pull request.

Profpatsch commented on lib/types-simple.nix:
```diff
@@ -0,0 +1,457 @@
+  # and it returns a more or less pretty string of errors.
+  prettyPrintErrors =
+    let
+      join = lib.foldl lib.concat [];
```
Good catch, will change.

Profpatsch commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -126,54 +124,77 @@ let
-  checkMeta = meta: if shouldCheckMeta then lib.remove null (lib.mapAttrsToList checkMetaAttr meta) else [];
+  checkMeta = meta:
+    if shouldCheckMeta
+    then lib.types-simple.checkType metaType meta
```
Nope, how would I measure it? Maybe cc @grahamc.

Right now the checks are not enabled by default, of course that would be a final goal.

Profpatsch commented on lib/types-simple.nix:
```diff
@@ -0,0 +1,457 @@
+
+  ## -- TYPES --
+
+  # the type with no inhabitants (kind of useless …)
```
Ah, it’s used in the tests for example: `(list void)` is a type with exactly one inhabitant, `[]`. One wouldn’t believe it, but void is a pretty handy type to have (the description is a bit tongue-in-cheek).

Profpatsch commented:
Here’s a (non-exhaustive) correspondence table.

module types | types-simple
-|-
`string` | n/a
`str` | `string`
`attrs` | `attrs any`
`list` | `list any`
`attrsOf` | `attrs`
`listOf` | `list`
`either l r` | `union [ l r ]`
`loaOf t` | `union [ (list t) (attrs t) ]`
`nullOr` | n/a
`submodule` | ~ `product`/`productOpt`
`enum` | ~ `sum` (but could make sense to have a separate type)
`ints.between` | n/a (stretch goal, maybe for arbitrary values, a flavor of dependent types)

Some of the module types come from the fact that they also contain merging logic and submodule logic, which complicates the base design and lead to the deprecation of some types.

Profpatsch committed:
lib/types-simple: remove defaults & small improvement

`defaults` originally only was a crutch to make products accept optional
attributes shallowly, but that has been replaced by `productOpt`, so we don’t
need it anymore.

Also a few other changes based on input by @aszlig.

Profpatsch commented:
Okay, rebase onto current master after the changes in `check-meta.nix`.

@Ericson2314 I noticed you create a lot of types in `lib/system/parse.nix`. I think changing the checks to `lib.types-simple` from this pull request would improve the type errors as well as how concise these types work (after adding an `enum` type). For now I’ve stubbed the `meta.platforms` fields out with `any`.

Profpatsch commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -133,45 +133,61 @@ let
+      # TODO: use the types from lib/systems/parse.nix
+      # any should be lib.systems.parsed.types.system
+      system = any;
+      platform = union [ string any ];
```
@Ericson2314 This is where I stub out the platform types at the moment.

Profpatsch committed:
lib/types-simple: add `restrict`

`restrict` is able to restrict the inhabitants of a type, meaning only a subset
of a type is allowed, the new type is “smaller”, but still gives nice type
errors if a nested value has the wrong type.

See the tests for some usage examples.

Profpatsch commented:
Update: `enum` and other restricted types are now possible using `restrict`.

edolstra commented on lib/default.nix:
```diff
@@ -31,6 +31,9 @@ let
     types = callLibs ./types.nix;
 
+    # type checking plain nix values
+    types-simple = callLibs ./types-simple.nix;
```
Please use camelCase for attribute names.

edolstra commented:
What is the memory / evaluation time impact of this?

edolstra commented on lib/tests/types-simple.nix:
```diff
@@ -0,0 +1,243 @@
+    };
+
+
+  # -- Products --
```
What are "products"? Seems like a pretty strange terminology.

Nadrieril replied:
`types-simple` is essentially defining [algebraic data types](https://en.wikipedia.org/wiki/Algebraic_data_type), like ML or Haskell have; "products" is the standard terminology for tuple- or record-like things in this setting.

edolstra commented:
Not sure whether this is a good idea. It seems to add *another* type system, separate from the one used by the NixOS module system. And the existence of the latter is already kind of unfortunate because Nix is a DSL for package/configuration management, so if we need types, we should add them to the language, *not* add them as a library feature, since that will always be slow and have clumsy syntax.

Also such a major expansion to the Nixpkgs library should really go through an RFC.

Ericson2314 commented:
@Profpatsch I was thinking about this thing when I worked on https://github.com/NixOS/nixpkgs/pull/34444; I think we can shoe-horn both type systems together (e.g. support flat and deep checking, and a structured attr thing that isn't submodule).

nbp commented:
@Profpatsch This new type definitions sound appealing as they are smaller than the NixOS one, because they to not have to deal with merging types nor extracting submodules.

However, It sounds painful and confusing to have both of them.

From what I understand, the only missing parts are the equivalent of the `product` without the requirement for submodules and Sum types. Still I do not see anything blocking an implementation of these for the existing type system.

Can you describe what problems did you faced with the existing type system, which prevented you from extending it?

@aszlig 
> it could also make sense to replace the type checking for the NixOS module system to use this instead (as it avoids the need to use tryEval for checking) on everything but submodules

There is no `tryEval` in the type, nor the module system. The check function will return if the type matches or not, and output an error message as it goes back with `addErrorContext`.

Profpatsch commented:
> Can you describe what problems did you faced with the existing type system, which prevented you from extending it?

There were multiple pain points, mainly stemming from the tight integration with the module system:

A while ago I tried implementing `sum` in `types.nix` and had a working check, but for the love of me could not figure out how to implement the required `merge`, `getSubModules`, `getSubOptions`, … functions and—more importantly—what I need them for exactly. I wasn’t even sure whether they could be implemented for that type at all.

More importantly though, I haven’t been able to figure out how to use `types.nix` to check whether a value `v` is of a type `t`, because one has to invoke the module system somehow to do that. [Looking at the current implementation of `check-meta.nix`](https://github.com/NixOS/nixpkgs/pull/37252/files#diff-096a3f0fd04e519accb1cde6c483e72bL172) I’m not the only one (except I was aware that e.g. `list.check` does not at all check the elements of a list.

Even if we were to implement a function `checkType = type: val:` for the `types.nix` types by invoking a function in `modules.nix` correctly, I wonder how much evaluation overhead the module merging and property discharging would cause. I can’t tell to be honest, because I haven’t been able to understand the implementation in `modules.nix`, even after revisiting it multiple times for multiple hours.

My strong preference would be to have a minimal type checking core with nice primitives and good structured, inspectable error messages (which I tried to create with `types-simple`) and use that to implement the type checks in `types.nix`/`modules.nix`/`options.nix`.

nbp commented:
Ok, correct me if I am wrong. The problem is a documentation issue, and you are hoping to have a simpler type-checking system, and later figure out how to use it in the module system. Correct?

You believe there is a potential for a better solution than the module system types, in which case I am hopeful that you might be right. However, what I know is that the `types-simple` system is quite similar to what we used to have the early days of NixOS modules:

https://github.com/NixOS/nixpkgs/blob/0e333688cea468a28516bf6935648c03ed62a7bb/lib/types.nix

So, to address this documentation issue, maybe you can send me one question at a time, and I will do my best to answer them.  Then we can later reuse this log of question to make an FAQ or a proper documentation.

This way, you would be able to answer if you can build a better type-system than what we have currently, and this we will probably help figure out what are the requirements for implementing a type system for Nix.

Does that sounds like a good approach to you? We only commit to get a better understanding and documentation of the current type system, before discussing whether one is better than the other, and if there is a potential for replacing the current one by a better one.

Profpatsch commented:
> So, to address this documentation issue, maybe you can send me one question at a time, and I will do my best to answer them. Then we can later reuse this log of question to make an FAQ or a proper documentation.

That seems like a good way to improve the module system.

##### Basic difference

> However, what I know is that the types-simple system is quite similar to what we used to have the early days of NixOS modules:

It is similar for scalar values, but it differs in that it 1) already contains merging logic and 2) doesn’t introduce a small set of axiomatic types.

The best example is the list or attributes of type, `loaOf`. It introduces a [quite complex merging logic](https://github.com/NixOS/nixpkgs/blob/0e333688cea468a28516bf6935648c03ed62a7bb/lib/types.nix#L114) to achieve the combination and keep good error messages, while in `types-simple` you’d just say `t: union [  (list t) (attr t) ]`. Union is an axiomatic type.
You could also do `t: union [ (list t) (product { foo = t; bar = string; }) ]` and get a complex nested type with maximally local error messages, no further work required.

The axiomatic types in `types-simple` are: `scalar` (non-nested types), `recursive` (lists and attrsets), `productOpt` (attrsets with differently typed (potentially optional) fields), `sum` (a tagged union) and `union` (an untagged union, which makes handling more difficult).

##### `checkType` for the module system

The question I’m most interested in (and that has the most value for uses like e.g. `check-meta.nix`):
How would an implementation of `checkType` look like in the current module system?

`checkType` in `types-simple` has the following properties:
* It takes a type `t` as first argument
* It takes a value `v` as second argument
* It returns `{}` if `v` is in the set of possible values of `t`
* It returns a localized error message for the parts of `v` that do not match their respective parts of `t` (which is a quite informal way to put it, I know)
  * I have [a big collection of tests](https://github.com/Profpatsch/nixpkgs/blob/85f28ffbb6089d0c0905606795ad27ac54118b1c/lib/tests/types-simple.nix) that give an overview of what errors are returned on mismatches
  * One example:
    ```
    checkType
      (attrs (attrs (list int)))
      { foo.bar = 42;
        foo.baz = [ 12 "foo" {} ]; }

    =>
    { foo.bar = { should = "list of integer"; val = 42; };
      foo.baz = {
        "1" = { should = "integer"; val = "foo"; };
        "2" = { should = "integer"; val = { }; };
     }; }
    ```

It should be noted that `checkType`’s [implementation only has to concern itself with the axiomatic types](https://github.com/Profpatsch/nixpkgs/blob/85f28ffbb6089d0c0905606795ad27ac54118b1c/lib/types-simple.nix#L92) and is thus pretty small and can easily be implemented again with some other type checking logic, should the need arise. For example a function `checkTypePredicate` can be written that just returns `true` or `false`. Or a checker that can annotate with some kind of position.

aszlig commented:
@nbp: The `tryEval` I was referring to is when you want to recursively check types via `mergeDefinitions`.

Ericson2314 commented:
Just wanted to say good luck, @nbp and @Profpatsch, and figuring somethings out!

Profpatsch committed:
lib/types-simple: enable “open” products

Open products can have additional fields in the actual value, which always
succeed in the type check (effectively default to `any`).

Profpatsch committed:
lib/types-simple: assert `req` and `opt` are disjunct

When defining a `productOpt` type, `req` and `opt` must not contain the same
fields, otherwise the type checking behaviour is undefined.

Profpatsch committed:
lib/generators: include name in toPretty for derivations

Profpatsch commented:
Update: `product`s can no also be `open`, meaning additional fields in the value are ignored/checked as `any`.
This is required to check e.g. a few fields of a derivation attrset, which can have arbitrarily many key/value pairs.

Profpatsch commented:
> I planned to make `bool` into a non-mergable bool and split other uses into `types.any` and `types.all` since `bool` has two different monoids.

Ah, the `Monoid` remark is enlightening, yes. Merging is really just `<>` (and defaults are `mempty`). Awesome!

oxij commented:
ping! Progress? Do you have an updated/rebased version of this to play with?

Profpatsch committed:
lib/types-simple: add path scalar type

Profpatsch committed:
lib/types-simple: test whether restrict works transitively

Profpatsch committed:
lib/types-simple: improve indentation a bit

Profpatsch commented:
@oxij I have just pushed a branch which does away with the in-file changes of `check-meta.nix` and gives you a type that can check derivation & meta fields: https://github.com/Profpatsch/nixpkgs/commit/7a2caa0d0a3cc99c02cc2b51293c852fb5992089

Profpatsch committed:
lib/types-simple: remove `names` attribute

`names` was originally intended to be able to match on the specific instance of
a type, but it breaks the abstraction in a way.
Since the function that used it was removed, we can remove the name attributes
as well.

oxij commented:
@Profpatsch awesome, thanks!

Shados commented:
@Profpatsch this is perhaps somewhat off-topic, but I read through `types-simple.nix`, and I found it rather tricky to understand the actual type checking code. Primarily, pulling the recursion out of `checkType` and into `cata` and `fmap` made the control flow egregiously difficult for me to follow, and `fmap` effectively changing the semantics of the nested types in-place was unexpected. In order to understand it, I ended up having to write a [version of it wihout either of those](https://gist.github.com/Shados/a51ab65b276310af22a29f461f255d46).

From the context, I gather that the design is based on ideas from category theory, and would (I assume) be easily comprehensible to people familiar with it. As a result, I am curious: are there other benefits in this specific instance to keeping the implementation so close to the abstract design?

Those comments notwithstanding, I *still* found it easier to understand than the existing module type system, and having a self-contained type system in place would be very useful.

Profpatsch commented:
> From the context, I gather that the design is based on ideas from category theory

No, not really, it’s from the classic bananas & barbed wire paper. You can find an explanation here: http://reasonablypolymorphic.com/blog/recursion-schemes/
It helps greatly when dealing with recursive data structures, by splitting open their recursion and just looking at one layer at a time.
You are right that it superficially complicates the code in an untyped setting, but in practice it is a great fit for a type checker which is intended to be expanded by users. If I expose the `alg` definition in `checkType` (which I should totally do), users can chain new transformations to each layer of recursion, e.g. to change error messages according to their needs.

mmahut commented:
Are there any updates on this pull request, please?

Profpatsch commented:
> Are there any updates on this pull request, please?

I still think it’s valuable, but I don’t know of any way out of the “we don’t want two type systems in nixpkgs” issue.
